feat: edit collection permission

Jarsen136 opened the pull request:
**Thank you for your contribution** to the [Koda - Generative Art Marketplace](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [x] Feature

## Needs QA check

- @kodadot/qa-guild please review

## Context

- [x] Closes #11148

## Screenshot 📸

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.
<img width="362" alt="image" src="https://github.com/user-attachments/assets/fe143ccd-21d1-4904-97f4-df8f7a9e7735" />

<img width="378" alt="image" src="https://github.com/user-attachments/assets/db2c1bb0-2f2a-4e31-8d1f-d751ad2a22d4" />

<img width="382" alt="image" src="https://github.com/user-attachments/assets/dab5f050-6107-4cb2-b144-5b64397eef7c" />

ref: https://wiki.polkadot.network/docs/learn-nft-pallets

Jarsen136 commented:
> Please use normal price input
> 

✅ Updated

preschian reviewed (approved):
nice, lgtm

hassnian commented on components/collection/EditModal.vue:
```diff
@@ -125,6 +125,56 @@
+
+            <div>
+              <div class="flex justify-between">
+                <p>{{ $t('mint.collection.permission.noPriceSet') }}</p>
```
after setting a price this did not change 

![CleanShot 2024-12-06 at 14 36 43@2x](https://github.com/user-attachments/assets/a7c33327-8faf-4ced-9fd3-9b527f64343f)

Jarsen136 replied:
refresh the page could help, but I will also add a refetch after signing

Jarsen136 replied:
Updated. This text will be changed after toggling.

hassnian commented on components/collection/HeroButtonEditCollection.vue:
```diff
@@ -67,14 +73,25 @@ const editCollection = async (collection: UpdateCollection) => {
   })
 }
+
+watchEffect(async () => {
+  const { apiInstance } = useApi()
+  const api = await apiInstance.value
+  const config = await api.query.nfts.collectionConfigOf(collectionId)
+  const mintSettings = (config.toHuman() as { mintSettings: CollectionMintSetting }).mintSettings
+
+  mintSettings.price = mintSettings.price?.replaceAll(',', '')
+  collectionPermissionSettings.value = mintSettings
+})
```
since this was added lets disable the `edit collection` button if `!collectionMetadata`

Jarsen136 replied:
added

hassnian commented on components/collection/EditModal.vue:
```diff
@@ -153,8 +203,11 @@ export type CollectionEditMetadata = {
+  mintingSettings: CollectionMintSetting
 }
 
+const COLLECTION_MINTING_TYPES_OPTIONS = ['Issuer', 'Public', 'HolderOf'].map(type => ({ value: type as CollectionMintSettingType, text: type }))
```
```suggestion
const COLLECTION_MINTING_TYPES_OPTIONS = (['Issuer', 'Public', 'HolderOf'] as CollectionMintSettingType[] ).map(type => ({ value: type, text: type }))
```

Jarsen136 replied:
updated

vikiival reviewed (changes requested):
~ incorrect impl

1. It does not make sense to have  issuer & price :) (will you pay for yourself?)

![Screenshot 2024-12-07 at 22 37 21](https://github.com/user-attachments/assets/fa6cba41-99a5-451b-8d81-6fd1ccb13106)

2. Setting public & set price > 0 means that you wont be able to mint on koda
2.1 Warning that setting public will allow anyone to mint into your collection 

3. Holder of is missing search input where user can select which address is eligible.
3.1 Warning that by setting type to holder of user wont be able to mint on koda

Jarsen136 commented:
> ~ incorrect impl
> 
> 1. It does not make sense to have  issuer & price :) (will you pay for yourself?)
> 2. Setting public & set price > 0 means that you wont be able to mint on koda
>    2.1 Warning that setting public will allow anyone to mint into your collection
> 3. Holder of is missing search input where user can select which address is eligible.
>    3.1 Warning that by setting type to holder of user wont be able to mint on koda
> 

Thanks for providing more details.

A question about the `holderOf` type. You're talking about the `address search input`. Shouldn't it be an input of `collection id`? With this collection id, we could know if the user is the holder of collection A and is eligible to mint nft on collection B.
<img width="1467" alt="image" src="https://github.com/user-attachments/assets/be327a2e-4209-43ca-ac8c-f6d657246f64">

vikiival commented:
> A question about the holderOf type. You're talking about the address search input. Shouldn't it be an input of collection id?

Yes, you are completely right, I apologize for my mistake. (I have edited the comment)

I meant 

```
 search input where user can select which collection is eligible
```

Why search input? 

Well if you take it from perspective of artist / @JustLuuuu they have no idea what `collection Id` is. (example https://koda.art/ahp/collection/258)
They do not know that this collection has id `258`, but they do know that collection is called `ANOMALY` ^-^

Jarsen136 commented:
> ~ incorrect impl
> 
> 1. It does not make sense to have  issuer & price :) (will you pay for yourself?)
> 2. Setting public & set price > 0 means that you wont be able to mint on koda
>    2.1 Warning that setting public will allow anyone to mint into your collection
> 3. Holder of is missing search input where user can select which collection is eligible.
>    3.1 Warning that by setting type to holder of user wont be able to mint on koda
> 



Updated

<img width="395" alt="image" src="https://github.com/user-attachments/assets/e2169243-2da3-4f46-a032-444cf9ab54a9">

<img width="379" alt="image" src="https://github.com/user-attachments/assets/7647ed50-2729-4f71-9181-ad0eaeb7ee11">
<img width="379" alt="image" src="https://github.com/user-attachments/assets/2e6e7336-6556-4032-86d9-544288f170f9">
<img width="393" alt="image" src="https://github.com/user-attachments/assets/a686b51e-2e36-48ea-b6b5-f9bb8c42ae3d">

hassnian commented:
no price has been set but i can still submit, if the switch is not meant to clickable we should hide it 

![CleanShot 2024-12-11 at 16 10 01@2x](https://github.com/user-attachments/assets/ff3b28e1-68b4-4baf-ae7a-18929931e5bb)

after pressing save changes 

![CleanShot 2024-12-11 at 16 11 44@2x](https://github.com/user-attachments/assets/4ba1ca5b-d2fe-4271-8e94-c5c6f0d29a51)

vikiival commented on components/collection/EditModal.vue:
```diff
@@ -172,6 +173,35 @@
+                class="mt-4"
+              >
+                <p class="mb-2">
+                  Holder of Collection
```
i18n

Jarsen136 replied:
updated

Jarsen136 commented:
> no price has been set but i can still submit, if the switch is not meant to be clickable we should hide it
> 

It's meant to disable the switch when `public` is selected. I have hidden the switch now

> 
> after pressing save changes
> 

In this case, I would disable the button.

Jarsen136 commented:
> [#11217 (review)](https://github.com/kodadot/nft-gallery/pull/11217#pullrequestreview-2486682215)
> 
> > 2. Setting public & set price > 0 means that you wont be able to mint on koda
> 
> I will try to rephrase as it is incorrectly implemented.
> 
> 
> * If you **set** price for public minting, you will not be able to mint in Koda.
> * Moreover anyone can mint into your collection

I'm not sure if I understand correctly. For the public type, should it allow the user to set zero price, or not?

Here is the current warning message implemented. Feel free to let me know how it is supposed to be.  @vikiival 
<img width="373" alt="image" src="https://github.com/user-attachments/assets/e054aab3-13c1-4f68-b39d-0c59c6d13da7" />
<img width="362" alt="image" src="https://github.com/user-attachments/assets/c5852894-6aee-4e2e-85a9-b2ad5f976b1d" />

Jarsen136 commented:
@vikiival Could you please take a look?

vikiival commented:
> @vikiival Could you please take a look?

Yup 
Can also @kodadot/internal-dev add review pls?

hassnian reviewed (changes requested):
1. holderOf - visual change

move the cross to the end of the container 

![CleanShot 2024-12-31 at 08 08 02@2x](https://github.com/user-attachments/assets/385654f1-620b-46d3-a0a3-c93bce04e7d7)

3. I can save changes even though no changes have been made with the selected mintType

**changing mintType should clear the state of that mintType** 

https://github.com/user-attachments/assets/32c31b3e-3882-4aa7-83fb-b10e6dff381a

3. holderOf , shouldn't the switch be reversed ?

active when price can be set
and off when not ?

![CleanShot 2024-12-31 at 08 15 48](https://github.com/user-attachments/assets/ad1cf0a4-a2d7-4b73-b8c9-51bb3e75f553)

4. public, error when saving changes 


https://github.com/user-attachments/assets/85c3a8e9-70df-4303-a982-4a20b2d3c811

5. holderOf

Ideally a loading state when searching would be nice 

https://github.com/user-attachments/assets/b1e939af-cd6b-46f0-a35c-825bc2d5d292

Jarsen136 commented:
> 1. holderOf - visual change
> 
> move the cross to the end of the container
> 

> 3. I can save changes even though no changes have been made with the selected mintType
> 
> **I think changing mintType should clear the state of that mintType**
> 
>  CleanShot.2024-12-31.at.08.10.27.mp4 
> 3. holderOf , shouldn't the switch be reversed ?
> 
> active when price can be set and off when not ?
> 
> 
> 4. public, error when saving changes
> 
>  CleanShot.2024-12-31.at.08.20.05.mp4 
> 5. holderOf
> 
> Ideally a loading state when searching would be nice
> 
>  CleanShot.2024-12-31.at.08.22.54.mp4

Thanks for all the review comments! I have updated them and implemented them all. @hassnian

exezbcz commented:
hey there, looking good, I would only change the logic of the input a bit

<img width="446" alt="image" src="https://github.com/user-attachments/assets/7d237ccc-3aa1-4743-8c48-87c7ec37be73" />



<img width="454" alt="image" src="https://github.com/user-attachments/assets/b4caab25-81b7-4d74-a6c4-7f5ad9926a4d" />


default is toggle off


Or do we want people to fill out specific price? Then it would make more sense to toggle it on by default.

Jarsen136 commented:
> hey there, looking good, I would only change the logic of the input a bit
> 
> <img alt="image" width="446" src="https://private-user-images.githubusercontent.com/90852205/399556921-7d237ccc-3aa1-4743-8c48-87c7ec37be73.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU3Mjk4NTYsIm5iZiI6MTczNTcyOTU1NiwicGF0aCI6Ii85MDg1MjIwNS8zOTk1NTY5MjEtN2QyMzdjY2MtM2FhMS00NzQzLThjNDgtODdjN2VjMzdiZTczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTAxVDExMDU1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRlYWUwN2UwYTRjZTRiMmNjMTk5NmJlNTQ0MTIxNDczNzRhMGIzOGQzNjNmNjcyMTQ3ZThlY2QyZDg4NzNlNGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9NSjUZbvgxWDla-p4DD55VUkazJbNLVxK361UcHTDwU"> <img alt="image" width="454" src="https://private-user-images.githubusercontent.com/90852205/399556840-b4caab25-81b7-4d74-a6c4-7f5ad9926a4d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU3Mjk4NTYsIm5iZiI6MTczNTcyOTU1NiwicGF0aCI6Ii85MDg1MjIwNS8zOTk1NTY4NDAtYjRjYWFiMjUtODFiNy00ZDc0LWE2YzQtN2Y1YWQ5OTI2YTRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTAxVDExMDU1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlNzU3NGFiNjliZTg1M2ExYzFkMzk5ZWVkNjhiOWM2OTVlNTE1ZWEwZDkyMDkyM2MyMjJkMGQ1OGE4OTM1OWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OH_yELvJfCKrQl05RNt7CvAF2Tzv146ws7L1DatfSvA">
> default is toggle off
> 
> Or do we want people to fill out specific price? Then it would make more sense to toggle it on by default.

It depends on which type the collection is.

There are three types of collection
public -> It has to set the price
issuer -> It can't set the price
holderOf -> It could have a price or not. In this case, it defaults off unless a price is already set on chain.

exezbcz commented:
@Jarsen136 this looks good, seems like i have seen older version or it was the holder of version that did not look right with the double negative, hassnians comment described it https://github.com/kodadot/nft-gallery/pull/11217#pullrequestreview-2526137124


**thank you for this overview** 


only change is the copy then: 
specify the price of the items -> Set specific price


as its a bit cleaner

Jarsen136 commented:
> @Jarsen136 this looks good, seems like i have seen older version or it was the holder of version that did not look right with the double negative, hassnians comment described it [#11217 (review)](https://github.com/kodadot/nft-gallery/pull/11217#pullrequestreview-2526137124)
> 
> **thank you for this overview**
> 
> only change is the copy then: specify the price of the items -> Set specific price
> 
> as its a bit cleaner

Thanks! The copy has been updated
<img width="366" alt="image" src="https://github.com/user-attachments/assets/d2d22e39-d785-4d4c-864e-5d8ceb53f30d" />

hassnian commented on components/collection/EditModal.vue:
```diff
@@ -186,10 +280,17 @@ const disabled = computed(() => {
+  const holderOfCollectionIdChanged = holderOfCollectionId.value !== originalHolderOfCollectionId.value
+  const invalidPublicCollection = selectedMintingType.value === 'Public' && !mintingPrice.value
+
+  const invalidHolderOfCollection = selectedMintingType.value === 'HolderOf' && !holderOfCollectionId.value
```
Create a computed , you can reuse it inside the template 

```suggestion
  const invalidHolderOfCollection = isHolderOfMintingType.value && !holderOfCollectionId.value
```

Jarsen136 replied:
updated

hassnian commented on components/collection/EditModal.vue:
```diff
@@ -186,10 +280,17 @@ const disabled = computed(() => {
+    || (!nameChanged.value && !descriptionChanged && !hasImageChanged.value && !hasBannerChanged && !hasMaxChanged && !mintTypeChanged.value && !mintPriceChanged.value && !holderOfCollectionIdChanged)
 })
 
+const permissionSettingWarningMessage = computed(() => selectedMintingType.value && permissionSettingCheckingMap[selectedMintingType.value] && permissionSettingCheckingMap[selectedMintingType.value]())
```
```suggestion
const permissionSettingWarningMessage = computed(() => selectedMintingType.value && permissionSettingCheckingMap[selectedMintingType.value]?.())
```

Jarsen136 replied:
updated

hassnian commented on components/collection/EditModal.vue:
```diff
@@ -215,6 +321,55 @@ watch(isModalActive, (value) => {
+}, {
+  immediate: true,
+})
+const mintTypeChangeHandlerMap = {
```
```suggestion
const mintTypeChangeHandlerMap: Record< CollectionMintSettingType, () => void> = {
```

Jarsen136 replied:
updated

hassnian commented on components/collection/EditModal.vue:
```diff
@@ -215,6 +321,55 @@ watch(isModalActive, (value) => {
+  },
+}
+
+const permissionSettingCheckingMap = {
```
Same here

Jarsen136 replied:
updated

hassnian commented on components/collection/SearchInput.vue:
```diff
@@ -0,0 +1,129 @@
+  immediate: true,
+})
+
+const handleSearch = async () => {
```
Let’s add a denounce

Jarsen136 replied:
updated

hassnian commented on composables/transaction/types.ts:
```diff
@@ -310,6 +310,13 @@ export interface ActionSetNftMetadata {
   errorMessage?: string
 }
 
+export type CollectionMintSettingType = 'Issuer' | 'Public' | 'HolderOf'
```
I would personally have it as an enum

Jarsen136 replied:
updated

hassnian reviewed (commented):
Nice! , small stuff

hassnian commented on components/collection/HeroButtonEditCollection.vue:
```diff
@@ -34,47 +35,83 @@ const { transaction, status, isLoading } = useTransaction()
 }
+
+watch(computed(() => collectionId), async () => {
+  const { apiInstance } = useApi()
+  const api = await apiInstance.value
+  const config = await api.query.nfts.collectionConfigOf(collectionId)
+  const mintSettings = (config.toHuman() as { mintSettings: {
+    price: string
+    mintType: CollectionMintSettingType
+    holderOf: string
+  } }).mintSettings
```
make it a util `getCollectionMintSettings`

```suggestion
  const mintSettings = getCollectionMintSettings(collectionId)
```

Jarsen136 replied:
updated

hassnian commented on components/collection/SearchInput.vue:
```diff
@@ -0,0 +1,130 @@
+<template>
+  <div class="relative w-full">
+    <input
+      v-if="!selectedCollection"
+      v-model="searchQuery"
+      type="text"
+      class="w-full h-12 p-2 border border-k-shade bg-background-color text-text-color focus:border-border-color outline-none"
+      :placeholder="$t('mint.collection.search')"
+      @input="handleSearch"
+    >
```
can NeoAutocomplete be used here ?

Jarsen136 replied:
It's not that straightforward to convert. I will keep it as it is now.
